[1.7.x] Fixed #24354 -- Prevented repointing of relations on superclasses when migrating a subclass's name change

MatthewWilkes opened the pull request:
When renaming a class the migrations framework follows all references backwards in order to repoint their foreign keys to the new location. However, the current implementation doesn't differentiate between relations by virtue of a direct relation to the subclass and those that it inherits from a superclass.

This change explicitly checks that the model that each found relation field belongs to is the same one being renamed. Without this, any project that has a renamed subclass will generate superfluous migrations for the parent class on every invocation of makemigrations, leaving incorrect migrations in eggs and spurious database records if not caught.

timgraham commented on tests/migrations/test_operations.py:
```diff
@@ -482,6 +482,22 @@ def test_rename_model_with_self_referential_fk(self):
+        self.assertNotIn(("test_rmwsc", "shetlandpony"), new_state.models)
+        self.assertIn(("test_rmwsc", "littlehorse"), new_state.models)
+        # RenameModel shouldn't repoint the superclass's relations, only local ones
+        self.assertEqual(project_state.models["test_rmwsc", "rider"].fields[1][1].rel.to, new_state.models["test_rmwsc", "rider"].fields[1][1].rel.to)
```
Please use this style to limit lines to 120 characters so we don't have to scroll to review it:
```
self.assertEqual(
    ...,
    ...
)
```

timgraham commented on tests/migrations/test_operations.py:
```diff
@@ -482,6 +482,22 @@ def test_rename_model_with_self_referential_fk(self):
+        """
+        Tests the RenameModel operation on model which has a superclass that has a foreign key
+        """
+
```
no newline

timgraham commented on tests/migrations/test_operations.py:
```diff
@@ -482,6 +482,22 @@ def test_rename_model_with_self_referential_fk(self):
 
+    def test_rename_model_with_superclass_fk(self):
+        """
+        Tests the RenameModel operation on model which has a superclass that has a foreign key
```
Please wrap docstrings at 79 characters and include a period at the end of the sentence.

timgraham commented:
Please squash commits and use a commit message like the PR title. I'll let a migrations expert give a final review, but this looks okay to me. It would be helpful if you could send a PR against master with the test and release note after this is merged.

MatthewWilkes commented:
Thanks for the review, I've made the requested changes and updated the PR. 

I'd be happy to submit another PR for the test against master, but I'm not sure about the changelog. Do you mean to keep the 1.7.x part of the changelog up to date, or to mention this in the 1.8/1.9 area? As my change doesn't require fixing on master it's just code coverage.

MarkusH commented on tests/migrations/test_operations.py:
```diff
@@ -482,6 +482,25 @@ def test_rename_model_with_self_referential_fk(self):
+        self.assertEqual(
+            project_state.models["test_rmwsc", "rider"].fields[1][1].rel.to,
+            new_state.models["test_rmwsc", "rider"].fields[1][1].rel.to
+        )
```
I'd like to have some actual tests for the underlying foreign keys: see the `self.assertFKExsits()` and `self.assertFKNotExists` usage in `test_rename_model_with_self_referential_fk()`

MarkusH commented on django/db/migrations/operations/models.py:
```diff
@@ -122,6 +122,10 @@ def state_forwards(self, app_label, state):
         del state.models[app_label, self.old_name.lower()]
         # Repoint the FKs and M2Ms pointing to us
         for related_object in (related_objects + related_m2m_objects):
+            if related_object.parent_model is not model:
```
I think we should go with `related_object.field.rel.to is not model` instead. It's more self explaining and should do the same.

MarkusH commented:
1.7.5 has been released. Can you update the patch with the annotations I made above, rebase on top of stable/1.7.x, and move the release note to `docs/releases/1.7.6.txt`, please.

MatthewWilkes commented:
Yup, absolutely.  Sorry, been busy launching sites, will get on this tomorrow or Friday. It hasn't fallen off my radar.

MarkusH commented:
Thanks, @MatthewWilkes .

timgraham commented:
Reminder that the time to fix this is running out. After 1.8 is released, 1.7 will only receive security fixes.

MatthewWilkes commented:
Apologies, I dropped the ball on this. I got distracted with Plone community stuff. I've updated this PR to target 1.7.7, added the requested FK tests and changed the condition to use `related_object.field.rel.to`

timgraham commented:
buildbot, test this please.

timgraham commented:
merged in d0607a7eee75121527cd03c89de82a222e298eae, thanks!
